### PR TITLE
Feature/separate mpi namespace

### LINKED ISSUE #182: Program SEGFAULTS when linking against both libmuscle_mpi and libmuscle
When a C++ or Fortran program is targeting the `mpi` version of libmuscle, but links against both the mpi and regular shared library, SEGFAULTS or other undefined behaviour may happen.

When linking with `-lmuscle -lmuscle_mpi` (in that order!), the following happens:
- When constructing the instance, the call signature (with `MPI_Comm` and `root` arguments) is only provided by the `mpi` library, so the muscle instance is created by the `mpi` enabled version of the library
- When calling other methods of the public API, both `libmuscle.so` and `libmuscle_mpi.so` provide an implementation. Since `-lmuscle` was provided first when linking, the non-MPI implementation provided by `libmuscle.so` is used! Since the non-MPI library has a different memory layout of the Instance object (it doesn't have the MPI members), this will likely trigger a SEGFAULT, undefined behaviour, or more subtle bugs.

Suggested solution:
- Implement the MPI constructor signature in the non-MPI version (and vice-versa), and make them throw a runtime error.

### PR DESCRIPTION
**Pull request by maarten-ic**
Note: depends on #209, which should be merged first. Suggestion to review this PR after #209 is merged.

Fixes #182

### CODE REVIEW

**Review by LourensVeen — APPROVED** [review]
Nice! I assume you tried some mixed-up linking scenarios and got useful error messages?

File: libmuscle/cpp/src/libmuscle/tests/test_api_guard.cpp [hunk 1]
```diff
@@ -5,8 +5,8 @@
 #include <vector>
 #include <set>
```
[the remark this answers is not in the document]
Comment (maarten-ic) [reply libmuscle/cpp/src/libmuscle/tests/test_api_guard.cpp:8]: Not really incidentally (as all headers that define the objects also import the namespace header), but agreed that it's good to add 😄

**Comment by maarten-ic** [comment]
> Nice! I assume you tried some mixed-up linking scenarios and got useful error messages?

Correct. I got it to either link and runs successfully, or get a linker error that symbols are not defined.

**Review by LourensVeen — APPROVED** [review]
Great!

**Pull request merged**